### PR TITLE
Implement callback functionality for Python 

### LINKED ISSUE #1068: Add callback interfaces for Python
Callback interfaces were implemented for Kotlin in #344 and for Swift in #1066.

This issue is for the work to bring callback interfaces to Python.

The Swift and Kotlin implementations of the bindings generator backends are very similar. The moving parts to be transcribed are:

 * `gen_swift/callbacks.rs`
 * `src/bindings/swift/templates/CallbackInterfaceTemplate.swift`
 * `src/bindings/swift/templates/CallbackInterfaceRuntime.swift`

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/UNIFFI-96)
┆Issue Number: UNIFFI-96

### PR DESCRIPTION
**Pull request by badboy**
We use the same pattern as for Swift and Kotlin:
We wrap a type, send over a simple callback function and do the
dispatching to the right instance method on the Python side again.

There's one catch:
`ctypes`' callback definition can't have the callback return a struct by value.
So we change the callback function slightly:

It returns an integer.
When positive it signals success.
It also gets an out pointer to a RustBuffer,
into which it writes the RustBuffer data.

Back on the Rust side a slot for the RustBuffer is set up,
passed in and afterwards used.
That's safe because an empty RustBuffer doesn't allocate, so there's
nothing to deallocate.

Sidenote:
Python's ctypes uses libffi.
There's were known issues with libffi's callbacks.
Current libffi versions should gracefully handle running under
SELinux. Previous versions might just crash when SELinux is enabled.
Another still-existing bug is about the interaction of libffi's callback
(de)allocation, when the application also uses `fork`.
We rely on functions not being GC'ed by Python and thus the callback never being deallocated,
thus not causing that issue for Uniffi-generated code.

Fixes #1068

---

NOTE: This is based on the branch of #1081. Only the last commit is new. It will be rebased once that's merged.

Regarding libffi issues see <https://cffi.readthedocs.io/en/latest/using.html?highlight=callbacks#callbacks-old-style>, <https://bugs.python.org/issue25653#msg402907> and <https://bugzilla.redhat.com/show_bug.cgi?id=1977410>

### CODE REVIEW

**Comment by badboy** [comment]
Rebased ... aaaaand I broke it.

**Comment by badboy** [comment]
Can I get another pair of eyes on this?

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt [hunk 1]
```diff
@@ -20,18 +20,25 @@ public interface {{ type_name }} {
...
 internal class {{ foreign_callback }} : ForeignCallback {
     @Suppress("TooGenericExceptionCaught")
-    override fun invoke(handle: Handle, method: Int, args: RustBuffer.ByValue): RustBuffer.ByValue {
+    override fun invoke(handle: Handle, method: Int, args: RustBuffer.ByValue, outBuf: RustBufferByReference): Int {
         val cb = {{ ffi_converter }}.lift(handle) ?: throw InternalException("No callback in handlemap; this is a Uniffi bug")
         return when (method) {
-            IDX_CALLBACK_FREE -> {{ ffi_converter }}.drop(handle)
+            IDX_CALLBACK_FREE -> {
+                {{ ffi_converter }}.drop(handle)
+                0
```
Comment (mhammond) [inline uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:28]: These magic values are a bit jarring - maybe drop a comment pointing at the docs for ForeignCallback?

File: uniffi_bindgen/src/bindings/python/gen_python/mod.rs [hunk 2]
```diff
@@ -77,24 +78,39 @@ impl<'a> PythonWrapper<'a> {
...
     pub fn members(&self) -> Vec<Box<dyn CodeDeclaration + 'a>> {
         let ci = self.ci;
-        vec![]
-            .into_iter()
-            .chain(ci.iter_enum_definitions().into_iter().map(|inner| {
+        vec![
+            Box::new(callback_interface::PythonCallbackInterfaceRuntime::new(ci))
+                as Box<dyn CodeDeclaration>,
+        ]
+        .into_iter()
```
Comment (mhammond) [inline uniffi_bindgen/src/bindings/python/gen_python/mod.rs:85]: mildly surprised by the indentation changes here? Is that cargo fmt? (I guess the fact the vec now has an element, the close `]` gets moved, and everything else follows?)

File: uniffi_bindgen/src/bindings/python/templates/CallbackInterfaceTemplate.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    # This should never happen, because an out of bounds method index won't
+    # ever be used. Once we can catch errors, we should return an InternalException.
+    # https://github.com/mozilla/uniffi-rs/issues/351
+    return -1
+
+# We need to keep this function reference alive:
+# if they get GC'd while in use then UniFFI internals could attempt to call a function
+# that is in freed memory.
+# That would be...uh...bad. Yeah, that's the word. Bad.
```
Comment (mhammond) [inline uniffi_bindgen/src/bindings/python/templates/CallbackInterfaceTemplate.py:72]: :)

File: uniffi_bindgen/src/bindings/python/templates/Helpers.py [hunk 4]
```diff
@@ -61,3 +61,7 @@ def rust_call_with_error(error_class, fn, *args):
     else:
         raise InternalError("Invalid RustCallStatus code: {}".format(
             call_status.code))
+
+# A function pointer for a callback as defined by UniFFI.
+# Rust definition `fn(handle: u64, method: u32, args: RustBuffer) -> RustBuffer`
```
Comment (mhammond) [inline uniffi_bindgen/src/bindings/python/templates/Helpers.py:66]: This comment has the old rust signature?

File: uniffi_bindgen/src/bindings/swift/templates/CallbackInterfaceTemplate.swift [hunk 5]
```diff
@@ -63,15 +63,18 @@ fileprivate let {{ foreign_callback }} : ForeignCallback =
         switch method {
             case IDX_CALLBACK_FREE:
                 {{ ffi_converter }}.drop(handle: handle)
-                return RustBuffer()
+                return 0
```
Comment (mhammond) [inline uniffi_bindgen/src/bindings/swift/templates/CallbackInterfaceTemplate.swift:66]: ditto here - comment pointing at where you doc'd the magic numbers?

**Review by mhammond — APPROVED** [review]
This looks great and is very thorough!

**Comment by thunderbiscuit** [comment]
I'm not sure if you are looking for community testing at this point (let me know if not and I'll just wait until it's ready for broader testing), but in case it's helpful to the development of this PR, I have not been able to build [our project using the PR](https://github.com/thunderbiscuit/bdk-ffi/tree/python-test-pr-1086), and this is the compilation error I'm getting:

```shell
Compiling uniffi_bindgen v0.15.2
   Compiling uniffi_bindgen v0.15.2 (https://github.com/badboy/uniffi-rs?branch=py/callbacks#5023329f)
   Compiling uniffi_build v0.15.2
   Compiling uniffi v0.15.2 (https://github.com/badboy/uniffi-rs?branch=py/callbacks#5023329f)
   Compiling bdk-ffi v0.1.0 (/Users/thunderbiscuit/repos/myforks/bdk-ffi)
   Compiling uniffi_macros v0.15.2
error[E0061]: this function takes 4 arguments but 3 arguments were supplied
    --> /Users/thunderbiscuit/repos/bdk-ffi/target/release/build/bdk-ffi-f355f45fd2d366e4/out/bdk.uniffi.rs:1148:18
     |
1148 |         unsafe { callback(self.handle, uniffi::IDX_CALLBACK_FREE, Default::default()) };
     |                  ^^^^^^^^ -----------  -------------------------  ------------------ supplied 3 arguments
     |                  |
     |                  expected 4 arguments

error[E0061]: this function takes 4 arguments but 3 arguments were supplied
    --> /Users/thunderbiscuit/repos/bdk-ffi/target/release/build/bdk-ffi-f355f45fd2d366e4/out/bdk.uniffi.rs:1166:33
     |
1166 |         let ret_rbuf = unsafe { callback(self.handle, 1, args_rbuf) };
     |                                 ^^^^^^^^ -----------  -  --------- supplied 3 arguments
     |                                 |
     |                                 expected 4 arguments

error[E0308]: mismatched types
    --> /Users/thunderbiscuit/repos/bdk-ffi/target/release/build/bdk-ffi-f355f45fd2d366e4/out/bdk.uniffi.rs:1167:37
     |
1167 |         uniffi::RustBuffer::destroy(ret_rbuf);
     |                                     ^^^^^^^^ expected struct `RustBuffer`, found `i32`

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
error: could not compile `bdk-ffi` due to 3 previous errors
```

**Comment by badboy** [comment]
> I'm not sure if you are looking for community testing at this point (let me know if not and I'll just wait until it's ready for broader testing), but in case it's helpful to the development of this PR, I have not been able to build [our project using the PR](https://github.com/thunderbiscuit/bdk-ffi/tree/python-test-pr-1086?rgh-link-date=2021-12-04T00%3A29%3A23Z), and this is the compilation error I'm getting:

Testing is always welcome. That looks like the scaffolding code wasn't regenerated with the code from this PR.

**Comment by thunderbiscuit** [comment]
Cheers! I'll keep an eye on this thread then!

As for compilation, I tried removing the `target` directory and start from scratch, but I'm getting the very same error when I attempt to cargo build.

Here is my `Cargo.toml`:
```toml
[dependencies]
bdk = { version = "0.13", features = ["all-keys", "use-esplora-ureq"] }
uniffi_macros = "=0.15.2"
uniffi = { git = "https://github.com/badboy/uniffi-rs", branch = "py/callbacks", features = ["builtin-bindgen"] }
thiserror = "1.0"

[build-dependencies]
uniffi_build = { version = "=0.15.2", features = [ "builtin-bindgen" ]}
```

**Comment by badboy** [comment]
> Cheers! I'll keep an eye on this thread then!
> 
> As for compilation, I tried removing the `target` directory and start from scratch, but I'm getting the very same error when I attempt to cargo build.
> 
> Here is my `Cargo.toml`:
> 
> ```toml
> [dependencies]
> bdk = { version = "0.13", features = ["all-keys", "use-esplora-ureq"] }
> uniffi_macros = "=0.15.2"
> uniffi = { git = "https://github.com/badboy/uniffi-rs", branch = "py/callbacks", features = ["builtin-bindgen"] }
> thiserror = "1.0"
> 
> [build-dependencies]
> uniffi_build = { version = "=0.15.2", features = [ "builtin-bindgen" ]}
> ```

You will need to use `{ git = "https://github.com/badboy/uniffi-rs", branch = "py/callbacks", features = ["builtin-bindgen"] }` for all `uniffi*` dependencies, as well as installing `uniffi-bindgen` (the cli tool, if you use it) from that branch as well, otherwise there will be a mismatch somewhere along the lines.

**Comment by thunderbiscuit** [comment]
It worked. I implemented a small part of our API and everything ran as expected.

Thank you!

**Pull request merged**